GameLogic/python_emulator.py

```diff
@@ -0,0 +1,531 @@
+    # Run on Init
+    #############################################################################
+
+    def __init__(self):
+
+        # Start Pygame
+        pygame.init()
+
+        # Display Constants
+        self.GRID_SIZE = 5  # scale entire pygame panel size
+        self.U_WIDTH, self.U_HEIGHT = 150, 60  # full panel proportions (20x19 dino)
+        self.FPS = 60  # frames per second
+        self.BG_COLOR = (255, 255, 255)  # white colour (background)
+        self.BLOCK_COLOR = (0, 0, 0)  # black colour (sprites)
+        self.UPDATE_WALK_FRAMES = 8  # how many frames to wait before dino switches legs
+        self.GND, self.REF, self.BIRD = (
+            4,
+            5,
+            20,
+        )  # proportions for how far apart the axis are seperated
+
+        # Game Constants
+        self.U_JUMP_HEIGHT = (
+            30  # dino is 20x19 in size, it should jump around its height, pixels
+        )
+        self.JUMP_TIME = 0.6  # full time to rise up and fall back down, seconds
+        self.CACTUS_CHANCE = (
+            0.4  # chance of cactus generation under this, float percent
+        )
+        self.MID_BIRD_CHANCE = (
+            0.5  # chance of mid bird generation between this and cactus, float percent
+        )
+        self.ADDITIONAL_COOLDOWN = (
+            0.5  # obstacles won't spawn so dino can't land, this is more wiggle room
+        )
+        self.ADDITIONAL_MOVING_SPEED = (
+            1  # baseline speed for obstacle movement speed, float magnification
+        )
+
+        # Run Initiation of Game States and Parameters
+        self.dino_array = []  # indicates which current dino sprite to display
+        self.dino_y = 0  # gives the verticle position of the dino
+        self.cactus_locs = 0  # gives the verticle axis of the cactus objects
+        self.mid_bird_locs = (
+            0  # gives the verticle axis of the middle flying bird objects
+        )
+        self.is_right_foot_down = (
+            True  # when True, the dino is in the right foot down walking position
+        )
+        self.is_jumping = (
+            False  # when True, the dino is mid jump (another jump cannot start)
+        )
+        self.is_ducking = False  # when True, the dino is shorter (won't hit mid birds)
+        self.start_game = False  # when False, shows the non-active resting game state
+        self.curr_walk = (
+            0  # used to count how many frames to wait before dino switches legs
+        )
+        self.jump_start_time = (
+            0  # used to count time while dino jumps for physics effect
+        )
+        self.cooldown_start_time = (
+            0  # used to count time so dino has enough time to land
+        )
+        self.cannot_spawn_object = False  # when True, no objects generated
+        self.setup_sprites()
+        self.setup_parameters()
+
+    def setup_sprites(self):
+        self.jump_array = np.array(
```

Comment:
I think `jump_array` and all these other attributes initialized in this function can be considered as constants since they are fixed values for `dino_array` or others to copy to? if that's the case we can also capitalize them to indicate they are consts.